### ISSUE TITLE
WaveNet example doesn't run

### ISSUE THREAD

**Issue by dhgrs** [chainer/chainer#8156]
* Conditions (using Docker image chainer/chainer:v6.3.0-python3)

Platform: Linux-4.15.0-58-generic-x86_64-with-Ubuntu-16.04-xenial
Chainer: 6.3.0
NumPy: 1.17.0
CuPy:
  CuPy Version          : 6.3.0
  CUDA Root             : /usr/local/cuda
  CUDA Build Version    : 9020
  CUDA Driver Version   : 10010
  CUDA Runtime Version  : 9020
  cuDNN Build Version   : 7600
  cuDNN Version         : 7602
  NCCL Build Version    : 2402
  NCCL Runtime Version  : 2402
iDeep: Not Available

* Code to reproduce
wavenet example in bca05f94b57a733d354fc361c29a79d783354deb

* Error messages, stack traces, or logs
```
Exception in main training loop:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/trainer.py", line 316, in run
    update()
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/updaters/standard_updater.py", line 175, in update
    self.update_core()
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/updaters/standard_updater.py", line 187, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/usr/local/lib/python3.5/dist-packages/chainer/optimizer.py", line 864, in update
    loss = lossfun(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/links/model/classifier.py", line 143, in forward
    self.y = self.predictor(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/mnt/vol21/dhigurashi/chainer/examples/wavenet/net.py", line 93, in forward
    encoded_condition = self.encoder(condition)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/mnt/vol21/dhigurashi/chainer/examples/wavenet/net.py", line 23, in forward
    x = F.relu(link(x))
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/links/connection/deconvolution_nd.py", line 151, in forward
    outsize=self.outsize, dilate=self.dilate, groups=self.groups)
  File "/usr/local/lib/python3.5/dist-packages/chainer/functions/connection/deconvolution_nd.py", line 404, in deconvolution_nd           ndim, stride, pad, outsize, dilate=dilate, groups=groups)
  File "/usr/local/lib/python3.5/dist-packages/chainer/functions/connection/deconvolution_nd.py", line 29, in __init__
    self.dilate = conv_nd.as_tuple(dilate, ndim)
  File "/usr/local/lib/python3.5/dist-packages/chainer/utils/conv_nd.py", line 13, in as_tuple
    assert len(x) == n
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "train.py", line 145, in <module>
    trainer.run()
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/usr/local/lib/python3.5/dist-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/trainer.py", line 316, in run
    update()
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/updaters/standard_updater.py", line 175, in update
    self.update_core()
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/updaters/standard_updater.py", line 187, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/usr/local/lib/python3.5/dist-packages/chainer/optimizer.py", line 864, in update
    loss = lossfun(*args, **kwds)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/links/model/classifier.py", line 143, in forward
    self.y = self.predictor(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/mnt/vol21/dhigurashi/chainer/examples/wavenet/net.py", line 93, in forward
    encoded_condition = self.encoder(condition)
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/mnt/vol21/dhigurashi/chainer/examples/wavenet/net.py", line 23, in forward
    x = F.relu(link(x))
  File "/usr/local/lib/python3.5/dist-packages/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/chainer/links/connection/deconvolution_nd.py", line 151, in forward
    outsize=self.outsize, dilate=self.dilate, groups=self.groups)
  File "/usr/local/lib/python3.5/dist-packages/chainer/functions/connection/deconvolution_nd.py", line 404, in deconvolution_nd           ndim, stride, pad, outsize, dilate=dilate, groups=groups)
  File "/usr/local/lib/python3.5/dist-packages/chainer/functions/connection/deconvolution_nd.py", line 29, in __init__
    self.dilate = conv_nd.as_tuple(dilate, ndim)
  File "/usr/local/lib/python3.5/dist-packages/chainer/utils/conv_nd.py", line 13, in as_tuple
    assert len(x) == n
AssertionError
```

The cause is https://github.com/chainer/chainer/pull/6285. It works when I revert the PR. We have to convert input dimension from 4D to 3D when we use Convolution1D instead of Convolution2D. In addition, this example has one or more processes based on 4D input like this.
https://github.com/chainer/chainer/blob/c4328a95a0db4adb8583f0a1d08ef03beabe2dfe/examples/wavenet/net.py#L49

**Comment by mitmul** [comment]
@dhgrs Are you working to resolve it? What's beyond my comprehension as I saw this was that the example tests seemed not to have reported any errors, but it turns out that the wavenet example is not tested at all with this: https://github.com/chainer/chainer-test/blob/master/test_example.sh
So, if you have already started fixing this, could you also add an example test with some right options?
Thanks.

**Comment by dhgrs** [comment]
I fixed this bug in #8157 but have not added any tests. So I will add some tests into the PR.

**Comment by dhgrs** [comment]
Sorry, test codes are in another repository so I will send another PR.
